mono/utils/mono-threads-posix.c

```diff
@@ -279,7 +280,31 @@ mono_native_thread_create (MonoNativeThreadId *tid, gpointer func, gpointer arg)
 void
 mono_threads_core_set_name (MonoNativeThreadId tid, const char *name)
 {
-#if defined (HAVE_PTHREAD_SETNAME_NP) && !defined (__MACH__)
+#ifdef __MACH__
+	static int (*pthread_setname_np_fn) (const char *);
+
+	/*
+	 * We can't set the thread name for other threads, but we can at least make
+	 * it work for threads that try to change their own name.
+	 */
+	if (tid != mono_native_thread_id_get ())
+		return;
+
+	/* The function is only available on OS X 10.6+, so find it at runtime. */
```

Comment:
http://www.mono-project.com/docs/about-mono/supported-platforms/osx/#introduction-to-mono-on-macos-x